v2/checksum.js

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+'use strict';
+
+var farm32 = require('farmhash').hash32;
+var crc32 = require('crc').crc32;
+
+var errors = require('./errors');
+var read = require('../lib/read');
+var write = require('../lib/write');
+
+module.exports = Checksum;
+
+// csumtype:1 (csum:4){0,1}
+function Checksum(type, val) {
+    if (!(this instanceof Checksum)) {
+        return new Checksum(type, val);
+    }
+    var self = this;
+    self.type = type;
+    self.val = val || 0;
+}
+
+Checksum.objOrType = function objOrType(arg) {
+    if (arg instanceof Checksum) {
```

Comment:
be wary of instanceof. If an application server is using tchannel@2.3 and they have a module doge-client using tchannel@2.2 you will run into a situation where there are two tchannels in node_modules and thus two Checksum prototypes.

In this situation `instanceof` fails to work.

The best known work arounds are detecting by string constants i.e. `arg && arg.type === 'Checksum'`. There are other type branding strategies that include globally caching uuids ( https://github.com/mmalecki/rust-result.js/blob/master/index.js#L4-L20 ) and check for the existance of `arg && arg.type === KNOWN_UUID`.

This may or may not be out of scope depending on usage of this method. Expect to have multiple copies of tchannel in production though because of how each client package will have it's own tchannel copy.